### ISSUE TITLE
Validation not working?

### ISSUE THREAD

**Issue by erotavlas** [SitePen/dmodel#9]
I created a Memory store (model) based on a simple Json Schema like this and tried to put an an invalid value, however the object was still created inside the store.

    <script>

        require(
            [
            'dojo/_base/declare',
            'dstore/Memory',
            'dmodel/extensions/jsonSchema'
            ], 
        function (declare, Memory, jsonSchema) {
            var myStore = new Memory({
            
                model: jsonSchema({
                    properties: {
                        someProperty: {
                            type: "number",
                            minimum: 0,
                            maximum: 10
                        },
                    }
                })
            });

            myStore.put({ id: 1, someProperty: -21 });

        });
    </script>

Next I tried to use the String Validator like this

    <script>

        require(
            [
            'dojo/_base/declare',
            'dstore/Memory',
            'dmodel/extensions/jsonSchema',
            'dmodel/validators/StringValidator'
            ], 
        function (declare, Memory, jsonSchema, StringValidator) {
            var myStore = new Memory({
            
                model: jsonSchema({
                    properties: {
                        someProperty: new StringValidator({
                            // must be at least 4 characters
                            minimumLength: 4,
                            // and max of 20 characters
                            maximumLength: 20,
                            // and only letters or numbers
                            pattern: /^\w+$/
                        }),
                    }
                })
            });

            myStore.put({ id: 1, someProperty: '#' });

        });
    </script>

Again the property was created in the store.  There were no validation errors.
Am I doing something wrong?  

My end goal is to build a model from json schema, then when data is entered into the model, to validate based on the constraints defined in the schema.  Data that does not match the constraints should not be put into the store. Is this even possible?

**Comment by arndbeissner** [comment]
The implementation was changed to use the property "Model" rather than "model". The documentation (README.md) doesn't reflect that - yet. Try "Model: jsonSchema(...)" instead and see if that works for you. In your example, the store will never invoke the dModel functionality.

**Comment by erotavlas** [comment]
I don't know what to do after I create the model.

            var myStore = new Memory({
                Model: jsonSchema({
                    properties: {
                        someProperty: {
                            type: "number",
                            minimum: 0,
                            maximum: 10
                        }
                    }
                })
            });

1) How do you change the value of a property while validating at the same time?
2) How do you get the value of a property?

I tried using set, and put, but I don't know the syntax so it complains that they are not functions, A working example would be appreciated.

**Comment by arndbeissner** [comment]
Re 1: see the readme.md near the end: "The dmodel/store/Validating module is a store mixin that adds functionality for validating any objects that are saved through put() or add()."

Re 2: with <object>.get("<propertyName>");

That said, I think some basics are wrong. You create a store (Memory) and pass a model. That means that any object that is added to the store will have the model added to its prototype chain. The same applies to any object you retrieve from the store.

All the methods you see in the readme are methods that are invoked *on an object from the store*, not on the store itself (or methods in invoked on a property). Like this:

var id=4711; // the id value of the object you retrieve
var myObject = myStore.get(id);
myObject.set("someProperty", 1);

**Comment by erotavlas** [comment]
Can the id be a string (for example a uuid)?

**Comment by arndbeissner** [comment]
Yes, a string is okay (actually that's what I use most of the time), but please look at the dStore documentation and tutorials. Docs for dModel are scarce, but quite good for dStore, and that's a dStore topic.

Also, please don't forget to close issue #10 (set is not a function). I think this is now clarified with this discussion.

**Comment by erotavlas** [comment]
Thank you for your assistance, I was able to get it working, however here are two observations so far.

1)  I was unable to use get(), for Memory store I think it is getSync() to get one of the objects. Is this right?

2) I'm not exactly sure if the validation is working or not.  For example for the property 'number' which according to the json schema should be a number of maximum value 10.  I tried putting two different values, a number greater than 10, and a value which was not a number (string)

An error is only produced in the first case, but not the second (even though the value fails to be set).  Is it supposed to generate an error if the data type doesn't match what is in the schema? Something like "must be of type number"?

```
    <script>

        require(
            [
            'dojo/_base/declare',
            'dstore/Memory',
            'dmodel/extensions/jsonSchema',
            'dmodel/validators/StringValidator',
            'dmodel/store/Validating',
            "dmodel/Model",
            "dojox/json/schema"
            ], 
        function (declare, Memory, jsonSchema, StringValidator, Validating, Model, DJS) {

            var validatingMemory = (declare([Memory, Validating]))({
                Model: jsonSchema({
                    properties: {
                        prime: {
                            type: 'boolean'
                        },
                        number: {
                            type: 'number',
                            minimum: 1,
                            maximum: 10
                        },
                        name: {
                            type: 'number',
                            required: true
                        }
                    }
                }),

                idProperty: "name"
            });


            validatingMemory.setData([
                { name: 'one', number: 1, prime: false, mappedTo: 'E' },
                { name: 'two', number: 2, prime: true, mappedTo: 'D' },
                { name: 'three', number: 3, prime: true, mappedTo: 'C' },
                { name: 'four', number: 4, even: true, prime: false, mappedTo: null },
                { name: 'five', number: 5, prime: true, mappedTo: 'A' }
            ]);

            console.log(validatingMemory.getSync("five").number);

            console.log(validatingMemory.get("five")); //undefined

            var testObject = validatingMemory.getSync("five");
            
            console.log(testObject.valueOf("name"));

            testObject.set("name","blah");

            console.log(testObject.valueOf("name"));

            var test1 = testObject.property("number"); //property object
            test1.validateOnSet = true;

            //egisters a listener for any changes to the value of this property
            //onlyFutureUpdates set to true if the callback should not be called for the current value (only future updates)
            test1.observe(
                function (newValue) {
                console.log("changed: " + newValue);
                },
                { onlyFutureUpdates: true }
            );

            //according to schema should only be able to put number between 1 and 10
            test1.put(55); //errors array contains one error - "must have a maximum value of 10"
            test1.put("5f5"); //errors array is undefined

        });
    </script>
```

**Comment by arndbeissner** [comment]
Little time ATM, just some hints:

- the property "name" in the schema is declared to be of type number. From your example it should probably be "string".  So you insert data into your store with setData that doesn't match the model. Not sure what that does when you access these items later.

- the extension jsonSchema.js delegates to validation code in a separate library json-schema. Might be a bug in that library, but fix the above issue first to make sure that doesn't mess up things.

**Comment by erotavlas** [comment]
Ok fixed the type on property name.  But it was not the problem. 

After looking at the json-schema library I noticed they do not handle type number.  According to the documentation for json-schema 'number' is one of the primitive types

> JSON Schema defines seven primitive types for JSON values:
> 
> array
> A JSON array.
> boolean
> A JSON boolean.
> integer
> A JSON number without a fraction or exponent part.
> number
> Any JSON number. Number includes integer.
> null
> The JSON null value.
> object
> A JSON object.
> string
> A JSON string.

But I don't see any handling for all the types including number

(this is the checkType function from validate.js of json-schema library

```
		// validate a value against a type definition
		function checkType(type,value){
			if(type){
				if(typeof type == 'string' && type != 'any' &&
						(type == 'null' ? value !== null : typeof value != type) &&
						!(value instanceof Array && type == 'array') &&
						!(value instanceof Date && type == 'date') &&
						!(type == 'integer' && value%1===0)){
					return [{property:path,message:(typeof value) + " value found, but a " + type + " is required"}];
				}
				if(type instanceof Array){
					var unionErrors=[];
					for(var j = 0; j < type.length; j++){ // a union type
						if(!(unionErrors=checkType(type[j],value)).length){
							break;
						}
					}
					if(unionErrors.length){
						return unionErrors;
					}
				}else if(typeof type == 'object'){
					var priorErrors = errors;
					errors = [];
					checkProp(value,type,path);
					var theseErrors = errors;
					errors = priorErrors;
					return theseErrors;
				}
			}
			return [];
		}
```
